### PR TITLE
Regression in loc.setitem losing mi names when df is empty

### LINKED ISSUE #46317: BUG: Difference in CSV roundtrip between 1.3.5 and 1.4.1
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
from io import StringIO
import pandas as pd

print(pd.__version__)
print("----------")

csv_file = StringIO("a,b,c,x")
csv_file.seek(0)

df = pd.read_csv(
    csv_file,
    index_col=["a", "b", "c"],
)

df.loc[(1,2,3)] = "foo"
print(df.to_csv())
```

### Issue Description

The example code above gives different outputs on Pandas 1.3.5 and 1.4.1.

On 1.3.5 I get
```none
1.3.5
----------
a,b,c,x
1,2,3,foo
```

On 1.4.1 I get
```none
1.4.1
----------
,,,x
1,2,3,foo
```

Note that for 1.4.1 the CSV columns corresponding to DataFrame index columns have no header.

### Expected Behavior

To me, the 1.3.5 behavior looks correct.

My use case is as follows: I have an CSV file that is initially empty, aside from the header line. I then have code that

1. Reads the CSV file
2. Updates the data
3. Re-writes the CSV file

This is broken in 1.4.1 because the index names get lost.

### Installed Versions

For 1.3.5:
<details>

INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-100-generic
Version          : #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.5
numpy            : 1.22.2
pytz             : 2019.3
dateutil         : 2.8.2
pip              : 22.0.3
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : 7.0.1
hypothesis       : None
sphinx           : 4.4.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3 (dt dec pq3 ext lo64)
jinja2           : 2.11.1
IPython          : 8.0.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.5.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 7.0.0
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : 1.4.31
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
numba            : None
</details>

For 1.4.1:
<details>
INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-100-generic
Version          : #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.5
numpy            : 1.22.2
pytz             : 2019.3
dateutil         : 2.8.2
pip              : 22.0.3
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : 7.0.1
hypothesis       : None
sphinx           : 4.4.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3 (dt dec pq3 ext lo64)
jinja2           : 2.11.1
IPython          : 8.0.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.5.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 7.0.0
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : 1.4.31
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #46317 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jreback** [comment]
thanks @phofl

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

File: doc/source/whatsnew/v1.4.2.rst [hunk 1]
```diff
@@ -17,6 +17,7 @@ Fixed regressions
 - Fixed regression in :meth:`DataFrame.drop` and :meth:`Series.drop` when :class:`Index` had extension dtype and duplicates (:issue:`45860`)
 - Fixed memory performance regression in :meth:`Series.fillna` when called on a :class:`DataFrame` column with ``inplace=True`` (:issue:`46149`)
 - Provided an alternative solution for passing custom Excel formats in :meth:`.Styler.to_excel`, which was a regression based on stricter CSS validation. Examples available in the documentation for :meth:`.Styler.format` (:issue:`46152`)
+- Fixed regression in :meth:`DataFrame.loc.__setitem__` losing :class:`MultiIndex` names if :class:`DataFrame`  was empty before (:issue:`46317`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.2.rst:20]: :meth:`DataFrame.loc.__setitem__` not a valid ref. will change in a release note tidy up

Comment (phofl) [reply doc/source/whatsnew/v1.4.2.rst:20]: What would be valid here? (For future references)

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.4.2.rst:20]: in 1.4.1 release notes https://pandas.pydata.org/pandas-docs/dev/whatsnew/v1.4.1.html#fixed-regressions , I changed it to

> Regression when setting values with [DataFrame.loc()](https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.DataFrame.loc.html#pandas.DataFrame.loc) losing [Index](https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.Index.html#pandas.Index) name if [DataFrame](https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.DataFrame.html#pandas.DataFrame) was empty before ([GH45621](https://github.com/pandas-dev/pandas/issues/45621))

will probably do similar

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.4.2.rst:20]: done in #46354

**Pull request merged**